src/y0/algorithm/conditional_independencies.py

```diff
@@ -0,0 +1,173 @@
+
+import networkx as nx
+from tqdm import tqdm
+from ananke.graphs import ADMG, SG
+
+from ..struct import ConditionalIndependency
+
+X = TypeVar('X')
+
+__all__ = [
+    'get_conditional_independencies',
+]
+
+
+def get_conditional_independencies(graph: ADMG,
+                                   *,
+                                   max_given: Optional[int] = None,
+                                   verbose: bool = False) -> Set[ConditionalIndependency]:
+    """Get the conditional independencies from the given ADMG.
+
+    :param graph: An acyclic directed mixed graph
+    :return: A set of conditional dependencies
+
+    .. seealso:: Original issue https://github.com/y0-causal-inference/y0/issues/24
+    """
+
+    separations = DSeparationJudgement.minimal(d_separations(graph, max_given=max_given, verbose=verbose))
+    independencies = {ConditionalIndependency.create(judgement.a, judgement.b, judgement.given)
+                      for judgement in separations}
+    return independencies
+
+
+class DSeparationJudgement:
+    """By default, acts like a boolean, but also caries evidence graph."""
+
+    def __init__(self, separated: bool, a, b, given, evidence: Optional[nx.Graph] = None):
+        """separated -- T/F judgement
+           a/b/given -- The question asked
+           evidence -- The end graph
+        """
+        self.separated = separated
+        self.a = a
+        self.b = b
+        self.given = given
+        self.evidence = evidence
+
+    def __bool__(self):
+        return self.separated
+
+    def __repr__(self):
+        return f"{repr(self.separated)} ('{self.a}' d-sep '{self.b}' given {self.given})"
+
+    def __eq__(self, other):
+        return self.separated == other
+
+    @classmethod
+    def minimal(cls, dseps, policy=None):
+        """Given some d-separations, reduces to a 'minimal' collection.
+
+        For indepdencies of the form A _||_ B | {C1, C2, ...} the minmal collection will:
+             * Have only one indepdency with the same A/B nodes.
+             * For sets of C nodes, replacement is made according to the 'policy' argument
+
+        'policy' -- Function from dseparation to representation suitable for sorting
+                    (used it is used as the 'key' function in python's 'sorted').
+                    The kept d-separation in an A/B pair will be the FIRST in the
+                    group sorted according to policy.
+
+        The default replacement policy is shortest set of givens & then lexicographic.
+        """
+        def _grouper(dsep):
+            """Returns a tuple of left & right side of a d-separation."""
+            return (dsep.a, dsep.b)
+
+        def _len_lex(dsep):
+            """Sort by length of givens & the lexicography a d-separation"""
+            return (len(dsep.given), ",".join(dsep.given))
+
+        policy = _len_lex if policy is None else policy
+
+        dseps = sorted(dseps, key=_grouper)
+        instances = {k: sorted(vs, key=policy)[0]
```

Comment:
`sorted(x, key=y)[0]` should be the same as `min(x, key=y)`

Reply:
Sure. Changed.